src/storage/heap_file.c

```diff
@@ -24743,5 +24748,265 @@ heap_nonheader_page_capacity ()
+
+      VPID_COPY (&prev_vpid, ((i == 0) ? (&null_vpid) : (&heap_pages_array[i - 1])));
+      VPID_COPY (&next_vpid, ((i == array_size - 1) ? (&null_vpid) : (&heap_pages_array[i + 1])));
+
+      error_code = heap_add_chain_links (thread_p, hfid, &heap_pages_array[i], &next_vpid, &prev_vpid,
+                                         &page_watcher, false, false);
+      if (error_code != NO_ERROR)
+        {
+          // This should never happen.
+          assert (false);
+          goto cleanup;
+        }
+    }
+
+  /**********************************************************/
+  /*        Now add the chain to the heap itself.           */
+  /**********************************************************/
+
+  // First get the heap header page.
+  error_code = heap_get_header_page (thread_p, hfid, &heap_hdr_vpid);
+  if (error_code != NO_ERROR)
+    {
+      ASSERT_ERROR ();
+      goto cleanup;
+    }
+
+  // Now get a watcher for the heap header page.
+  error_code = heap_get_page_with_watcher (thread_p, &heap_hdr_vpid, &heap_header_watcher);
+  if (error_code != NO_ERROR)
+    {
+      ASSERT_ERROR ();
+      goto cleanup;
+    }
+
+  // Get the heap header.
+  heap_hdr = heap_get_header_stats_ptr (thread_p, heap_header_watcher.pgptr);
+  if (heap_hdr == NULL)
+    {
+      assert (false);
+      error_code = ER_FAILED;
+      goto cleanup;
+    }
+
+  // Save for logging.
+  heap_hdr_prev = *heap_hdr;
+
+  // Get the last page of the heap.
+  error_code = heap_get_last_page (thread_p, hfid, heap_hdr, NULL, &heap_last_page_vpid, &heap_last_page_watcher);
+  if (error_code != NO_ERROR)
+    {
+      ASSERT_ERROR ();
+      goto cleanup;
+    }
+
+  // Add new links to the first page of the chain.
+  error_code = heap_add_chain_links (thread_p, hfid, &heap_pages_array[0], NULL, &heap_last_page_vpid,
+                                     &page_watcher, true, false);
+  if (error_code != NO_ERROR)
+    {
+      ASSERT_ERROR ();
+      goto cleanup;
+    }
+
+  // Add new links to the last page of the heap.
+  error_code = heap_add_chain_links (thread_p, hfid, &heap_last_page_vpid, &heap_pages_array[0], NULL,
+                                     &heap_last_page_watcher, true, true);
+  if (error_code != NO_ERROR)
+    {
+      ASSERT_ERROR ();
+      goto cleanup;
+    }
+
+  // Now update the last page of the heap header.
+
+  heap_hdr->estimates.last_vpid = heap_pages_array[array_size - 1];
+  heap_hdr->estimates.num_pages += (int) array_size;
+
+  // Log this change.
+  addr.pgptr = heap_header_watcher.pgptr;
+  addr.vfid = &hfid->vfid;
+  addr.offset = HEAP_HEADER_AND_CHAIN_SLOTID;
+
+  log_append_undoredo_data (thread_p, RVHF_STATS, &addr, sizeof(HEAP_HDR_STATS), sizeof (HEAP_HDR_STATS), 
+                            &heap_hdr_prev, heap_hdr);
+
+  // Set the page as dirty.
+  pgbuf_set_dirty (thread_p, heap_header_watcher.pgptr, DONT_FREE);
+
+   // Now we can commit the sysop.
+  log_sysop_commit (thread_p);
+
+cleanup:
```

Comment:
I'd like to add this to make it sure:
```c
if (error_code != NO_ERROR)
  {
    log_sysop_abort (thread_p);
  }
```

We may change with above commit as follows:
```c
if (error_code != NO_ERROR)
  {
    log_sysop_abort (thread_p);
  }
else
  {
    log_sysop_commit (thread_p);
  }
```